File: agentlens/sdk-node/src/processors.ts
// Pipeline de transformación de spans antes de exportar (paridad con
// sdk-python/processors.py). Orden: normalización de convenciones -> redacción
// de PII -> externalización de payloads.

import type { Attributes } from "@opentelemetry/api";
import type { ReadableSpan, SpanExporter } from "@opentelemetry/sdk-trace-base";
import type { ExportResult } from "@opentelemetry/core";

import type { AgentLensConfig } from "./config.js";
import { normalizeAttributes } from "./conventions.js";
import {
  InMemoryPayloadStore,
  NoopPayloadStore,
  PayloadStore,
  discardContent,
  externalizeAttributes,
} from "./payloads.js";
import { Redactor } from "./redaction.js";

/** Envuelve un span exponiendo unos atributos transformados, preservando el resto. */
function spanWithAttributes(span: ReadableSpan, attributes: Attributes): ReadableSpan {
  return new Proxy(span, {
    get(target, prop, receiver) {
      if (prop === "attributes") return attributes;
      const value = Reflect.get(target, prop, receiver);
      return typeof value === "function" ? value.bind(target) : value;
    },
  });
}

export class AgentLensSpanExporter implements SpanExporter {
  private readonly redactor?: Redactor;
  private readonly store: PayloadStore;

  constructor(
    private readonly inner: SpanExporter,
    private readonly config: AgentLensConfig,
    opts: { redactor?: Redactor; store?: PayloadStore } = {},
  ) {
    this.redactor = opts.redactor ?? (config.redactPii ? new Redactor() : undefined);
    this.store =
      opts.store ??
      (config.payloadMode === "reference" ? new InMemoryPayloadStore() : new NoopPayloadStore());
  }

  private transform(span: ReadableSpan): ReadableSpan {
    let attrs: Attributes = normalizeAttributes(span.attributes);
    if (this.redactor) {
      attrs = this.redactor.redactAttributes(attrs);
    }
    if (this.config.payloadMode === "reference") {
      attrs = externalizeAttributes(
        attrs,
        this.store,
        this.config.contentAttrs,
        this.config.payloadThresholdBytes,
      );
    } else if (this.config.payloadMode === "none") {
      attrs = discardContent(attrs, this.config.contentAttrs);
    }
    return spanWithAttributes(span, attrs);
  }

  export(spans: ReadableSpan[], resultCallback: (result: ExportResult) => void): void {
    this.inner.export(spans.map((s) => this.transform(s)), resultCallback);
  }

  shutdown(): Promise<void> {
    return this.inner.shutdown();
  }

  forceFlush(): Promise<void> {
    return this.inner.forceFlush?.() ?? Promise.resolve();
  }
}

File: Agentlens/agentlens/sdk-python/src/agentlens/processors.py
"""Pipeline de transformación de spans antes de exportar.

``AgentLensSpanExporter`` envuelve al exporter real (OTLP) y, por cada span,
aplica en este orden:

  1. Normalización de convenciones (alias legacy/variantes -> esquema canónico)
  2. Redacción de PII   (para que no llegue ni al almacén de payloads)
  3. Externalización de payloads grandes

Como los atributos de un span finalizado son inmutables (``BoundedAttributes``
lanza ``TypeError`` al asignar), reconstruimos un ``ReadableSpan`` nuevo con los
atributos transformados. Este enfoque es independiente del orden de los
procesadores y compatible con el exporter OTLP real (verificado).
"""
from __future__ import annotations

from typing import Optional, Sequence

from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult

from .config import AgentLensConfig
from .conventions import normalize_attributes
from .payloads import (
    LocalFilePayloadStore,
    NoopPayloadStore,
    PayloadStore,
    discard_content,
    externalize_attributes,
)
from .redaction import Redactor


class AgentLensSpanExporter(SpanExporter):
    def __init__(
        self,
        inner: SpanExporter,
        config: AgentLensConfig,
        redactor: Optional[Redactor] = None,
        store: Optional[PayloadStore] = None,
    ):
        self._inner = inner
        self._config = config
        self._redactor = redactor if (redactor or config.redact_pii) else None
        if config.redact_pii and self._redactor is None:
            self._redactor = Redactor()
        if store is not None:
            self._store = store
        elif config.payload_mode == "reference":
            self._store = LocalFilePayloadStore()
        else:
            self._store = NoopPayloadStore()

    def _transform(self, span: ReadableSpan) -> ReadableSpan:
        attrs = dict(span.attributes or {})

        # Capa adaptadora de convenciones: homogeneiza el esquema venga de donde
        # venga el span (helpers propios o auto-instrumentación de terceros)
        # antes de redactar/externalizar, que ya trabajan contra claves canónicas.
        attrs = normalize_attributes(attrs)

        if self._redactor is not None:
            attrs = self._redactor.redact_attributes(attrs)

        if self._config.payload_mode == "reference":
            attrs = externalize_attributes(
                attrs, self._store, self._config.content_attrs,
                self._config.payload_threshold_bytes,
            )
        elif self._config.payload_mode == "none":
            attrs = discard_content(attrs, self._config.content_attrs)

        return ReadableSpan(
            name=span.name,
            context=span.context,
            parent=span.parent,
            resource=span.resource,
            attributes=attrs,
            events=span.events,
            links=span.links,
            kind=span.kind,
            instrumentation_scope=span.instrumentation_scope,
            status=span.status,
            start_time=span.start_time,
            end_time=span.end_time,
        )

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        transformed = [self._transform(s) for s in spans]
        return self._inner.export(transformed)

    def shutdown(self) -> None:
        self._inner.shutdown()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return self._inner.force_flush(timeout_millis)
